pkg/daemon/update.go

```diff
@@ -184,44 +184,21 @@ func (dn *Daemon) reconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) *stri
 			// check if the prior config is empty and that this is the first time running.
 			// if so, the SSHKey from the cluster config and user "core" must be added to machine config.
 			if len(oldIgn.Passwd.Users) == 0 && len(newIgn.Passwd.Users) == 1 {
-				if newIgn.Passwd.Users[0].Name == "core" && len(newIgn.Passwd.Users[0].SSHAuthorizedKeys) > 0 {
-					glog.Info("SSH Keys reconcilable")
-				} else {
-					msg := "Ignition passwd user section contains unsupported changes"
+				msg := verifyUserFields(newIgn.Passwd.Users[0])
+				if msg != "" {
 					return &msg
 				}
-			} else if len(oldIgn.Passwd.Users) == 1 && len(newIgn.Passwd.Users) == 1 {
-				// the length of oldIgn.Passwd.Users should always be 1. We do not supported adding
-				// users outside of the initial addition of the user "core".
-				if oldIgn.Passwd.Users[0].Name == "core" && newIgn.Passwd.Users[0].Name == "core" && len(newIgn.Passwd.Users[0].SSHAuthorizedKeys) > 0 {
-					// we will only do an update if the new config contains has the name "core" and at least one ssh key
-					if !reflect.DeepEqual(oldIgn.Passwd.Users[0].SSHAuthorizedKeys, newIgn.Passwd.Users[0].SSHAuthorizedKeys) {
-						// keys don't match, let's create 2 temp Users
-						tempOld := oldIgn.Passwd.Users[0]
-						tempNew := newIgn.Passwd.Users[0]
-						// Let's set both Users' SSHKeys to match
-						tempOld.SSHAuthorizedKeys = tempNew.SSHAuthorizedKeys
-						// Now compare the two Users controlling for the SSHAuthorizedKey diff
-						if reflect.DeepEqual(tempOld, tempNew) {
-							// only diff between the two Passwds was the SSH key
-							glog.Info("SSH Keys reconcilable")
-						} else {
-							// there are differences other than just the SSH key
-							msg := "Ignition passwd user section contains unsupported changes"
-							return &msg
-						}
-					} else {
-						msg := "Ignition passwd user section contains unsupported changes"
+			} else if len(oldIgn.Passwd.Users) > 0 && len(newIgn.Passwd.Users) >= 1 {
+				for _, user := range newIgn.Passwd.Users {
+					if user.Name != "core" {
```

Comment:
It'd probably be clearer to explicitly error out if `len(newIgn.Passwd.Users) > 1` since we only support the "core" user, and there can only be one user with a given name so something like:

```
if len(newIgn.Passwd.Users) > 1 {
   msg := "Adding multiple users is not supported"
   return &msg
}
```

If we do that *first*, every following conditional can then be sure the length is either 0 or 1.

Reply:
Basically all the code after that could actually operate on a `*User` which can be `nil` (in Rust terms `Option<&User>`) and we don't need to think about arrays of users.

IOW after the above code, something like:

```
newUser := nil
oldUser := nil
if len(newIgn.Passwd.Users) > 0 {
    newUser = &newIgn.Passwd.Users[0];
}
if len(newIgn.Passwd.Users) > 0 {
    oldUser = &oldIgn.Passwd.Users[0];
}
```

And then instead of doing `newUsers[0].Name` it'd just be `newUser.Name`.

Reply:
You would think this would work and this suggestion is closer to how this PR was initially structured _before_ I discovered Ignition's quirks :sob:  

Short explanation: the way Ignition works the newUsers array is actually the compilation of all the Users from all of the machineconfigs in `oc get machineconfigs`. What ignition does is only _apply_ the last one. So that the newUsers array will likely always be greater than 1, except for the first time this runs when getting the initial SSH Key from the installer.